### PR TITLE
Merge the `daemon` branch

### PR DESCRIPTION
**Pull request by riccardomurri**
Review request from @arcimboldo

### CODE REVIEW

File: gc3apps/irm.uzh.ch/gbeastcli.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import os
+
+if len(sys.argv) < 2:
+    print("Usage: %s <path-to-socket>" % sys.argv[0])
+    sys.exit(1)
+
+path=os.path.abspath(sys.argv[1])
+sys.argv[1] = '--'
+# Seems that the client program also get commands via arguments
+ctl = oi.CtlProgram('cli', address='ipc://%s' % path)
```
Comment (riccardomurri) [inline gc3apps/irm.uzh.ch/gbeastcli.py]: Are we still using `oi`?  If not, this file should be deleted.

File: gc3apps/irm.uzh.ch/gbeastdaemon.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    It also check, through inotify, if new files are added to the
+    directory, in that case, a new application is added to the
+    session.
+
+    """
+    version = '0.1'
+
+    def setup_options(self):
+        """Script to parallelize execution of BEAST over multiple input
+        files. Takes a directory containing XML files, parse"""
```
Comment (riccardomurri) [inline gc3apps/irm.uzh.ch/gbeastdaemon.py]: Unfinished sentence?

File: gc3apps/irm.uzh.ch/gbeastdaemon.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        self.add_param('--cores', default=1, type=int, help="Amount of cores to use. Default: %(default)s")
+
+
+    def new_tasks(self, extra, epath=None, emask=0):
+        # Scan of the directory is performed in `before_main_loop`,
+        # here we only need to ensure the fake always-running task is
+        # created.
+
+        if not epath or not epath.endswith('.xml'):
+            gc3libs.log.info("Ignoring file %s since it doesn't end in .xml" % epath)
```
Comment (riccardomurri) [inline gc3apps/irm.uzh.ch/gbeastdaemon.py]: If `epath` is `None` then this ends up printing the message:

    Ignoring file None since it doesn't end in .xml

while technically true, I find the message rather weird.  Can't we have two messages
for the two cases?

File: gc3apps/irm.uzh.ch/gbeastdaemon.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            # them.
+
+            # FIXME: What should we do when a file is *updated*?
+            # If you touch an existing file, this will trigger
+            # the creation of a new instance. You could check if
+            # fname is not in
+            #
+            #    [app.path for app in self.session if 'path' in app]
+            #
+            # but it might not be what you want. Maybe a command
```
Comment (riccardomurri) [inline gc3apps/irm.uzh.ch/gbeastdaemon.py]: "it might not be what you want" -> what do you have in mind?  I would *verbosely ignore* updates by default.

Anyway, whatever we do it should be clearly stated in the program's help text, and it's not ATM as far as I can see.

File: gc3apps/irm.uzh.ch/gbeastdaemon.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    def setup_options(self):
+        """Script to parallelize execution of BEAST over multiple input
+        files. Takes a directory containing XML files, parse"""
+
+        self.add_param('--beast1', default='/apps/BEASTv1.8.2/lib/beast.jar', help='Path to BEAST v1 jar file')
+        self.add_param('--beast2', default='/apps/BEASTv2.3.2/lib/beast.jar', help='Path to BEAST v2 jar file')
+        self.add_param('--cores', default=1, type=int, help="Amount of cores to use. Default: %(default)s")
+
+
+    def new_tasks(self, extra, epath=None, emask=0):
```
Comment (riccardomurri) [inline gc3apps/irm.uzh.ch/gbeastdaemon.py]: `emask` is unused AFAICS

File: gc3libs/tests/scripts/daemoncli.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import os
+
+if len(sys.argv) < 2:
+    print("Usage: %s <path-to-socket>" % sys.argv[0])
+    sys.exit(1)
+
+path=os.path.abspath(sys.argv[1])
+del sys.argv[1]
+# Seems that the client program also get commands via arguments
+ctl = oi.CtlProgram('cli', address='ipc://%s' % path)
```
Comment (riccardomurri) [inline gc3libs/tests/scripts/daemoncli.py]: Is `oi` still being used? I do not see it in the daemon sources.

File: gc3libs/tests/scripts/simpledaemon.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+import os
+import gc3libs
+from gc3libs.cmdline import SessionBasedDaemon
+import inotifyx
+
+class SimpleDaemon(SessionBasedDaemon):
+    """Simple daemon"""
+    version = '1.0'
+    application = 'simpledaemon'
```
Comment (riccardomurri) [inline gc3libs/tests/scripts/simpledaemon.py]: Not needed - this is not an `Application` subclass

File: gc3apps/irm.uzh.ch/gbeastdaemon.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+#
+__docformat__ = 'reStructuredText'
+__version__ = '$Revision$'
+
+
+import gc3libs
+from gc3libs.cmdline import SessionBasedDaemon
```
Comment (riccardomurri) [inline gc3apps/irm.uzh.ch/gbeastdaemon.py]: I cannot see this class anywhere in the branch code -- maybe you forgot to `git add` this file?

**Comment by riccardomurri** [comment]
Looks good so far, but the `SessionBasedDaemon` class is missing ;-)

**Comment by riccardomurri** [comment]
Sorry, it's GitHub that's hiding the changes from the "diff" view:

    1,091 additions, 488 deletions not shown because the diff is too large. Please use a local Git client to view these changes.

**Comment by riccardomurri** [comment]
In cmdline.py:

>         FIXME: While on a SessionBasedScript this method is called
>         *after* processing all the jobs, on a SessionBasedDaemon it is
>         called *before*.

Why cannot this be fixed now? (I.e., before merging)

**Comment by riccardomurri** [comment]
In cmdline.py:

>         # Start XMLRPC server
>         self.server = sxmlrpc.SimpleXMLRPCServer(("localhost", 0),
>                                                  logRequests=False)

Why hard-code `localhost` here?

**Comment by riccardomurri** [comment]
In cmdline.py:

>     def list_jobs(self, opts=None):
>         """usage: list [detail]\n\nList jobs"""

Why are those literal `\n` needed?

**Comment by riccardomurri** [comment]
In cmdline.py:

>         if opts and 'details'.startswith(opts):
>             def print_app_table(app, indent, recursive):

Please do not create local functions unless this is really needed to
capture local variables (e.g., like you did in the `killme()` function
within `terminate()`)

A local function creates a closure, which means the function must be
created by Python at runtime, and each time you execute the parent
function you get a new instantiation of the local function.  Unless a
closure is really needed, this is a waste of time and memory.

Bottom line: use a class' staticmethod or a top-level private function
whenever possible.

**Comment by riccardomurri** [comment]
In cmdline.py, function `stat_jobs()`:

>         stats = defaultdict(int)
>         for job in self.parent.session:
>             stats[job.execution.state] += 1
>         return str.join(' ', ["%s:%s" % x for x in stats.items()])

Isn't this already available as `Engine.stats()`, i.e., `self.controller.stats()`?

**Comment by riccardomurri** [comment]
In cmdline.py:

>             self.commthread._Thread__stop()
>             self.commthread._Thread__delete()

Ugh! Is calling a private method of the `Thread` class the only way of killing it?

**Comment by riccardomurri** [comment]
In cmdline.py:

>     def setup(self):
>         _SessionBasedCommand.setup(self)
>         # change default for the core/memory/walltime options

Comment is out of date and does not refer to the following code.

**Comment by riccardomurri** [comment]
In cmdline.py:

>         self.add_param('--comm', action='store_true',
>                        help='Enable communication socket')

Maybe `--listen` is a better name?  Could also be extended to take an
optional argument specifying which IP address and port to listen on.

**Comment by riccardomurri** [comment]
In cmdline.py:

>         self.add_param('--client',
>                        nargs='+',
>                        metavar="ARGS",

I'm not sure I like this to be implemented as an *option*; the problem
is that it is then legal to combine it with other options, and the
result is (a priori) undefined: what should the effect of `daemon
--notify-state FOO --client BAR` be?

Wouldn't it be better to have the `SessionDaemon` class take a
mandatory first argument: *server* or *client* and then the rest of
the command line is parsed accordingly::

        daemon server --notify-state FOO

        daemon client /path/to/sock bar baz

Doesn't solve the "grammar" problem (unless we switch, e.g., to click
for parsing command-line options) but at least one can state in the
help text "this option only works in *server* mode" or "this option
only works in *client* mode".

**Comment by riccardomurri** [comment]
In cmdline.py:

>         ### FIXME: first half of this function is copied from
>         ### _SessionBasedCommand().

Why not factor common code into a separate method of _SessionBasedCommand?

**Comment by riccardomurri** [comment]
In cmdline.py:

>             self.log.debug("Create working directory %s",

should be:

>             self.log.debug("Creating working directory `%s`  ...",

or even better:

>             self.log.debug("Working directory `%s` does not exist, creating it ...",

**Comment by riccardomurri** [comment]
In cmdline.py:

>             os.mkdir(self.params.working_dir)

You might want to use `os.makedirs` instead, to create missing parent directories.

**Comment by riccardomurri** [comment]
In cmdline.py:

>                 self.log.warning("Create non-existent inbox directory %s",
>                                  inbox)
>                 os.mkdir(inbox)

See the two comments above; same applies here.

**Comment by riccardomurri** [comment]
In cmdline.py:

>             if istate not in inotifyx.constants.keys():

better:

>             if istate not in inotifyx.constants:

Reason: it's shorter and works with Python 3 as well.

**Comment by riccardomurri** [comment]
In cmdline.py:

>     def __setup_inotify(self):
>         # Setup inotify on inbox directories
>         self.inotify_fds = {}
>         for inbox in self.params.inbox:
>             ifd = inotifyx.init()
>             self.inotify_fds[inbox] = ifd
>             inotifyx.add_watch(ifd, inbox, self.inotify_event_mask)

Why is a different inotify object needed *per directory*?
Can't you just `add_watch()` all directories on a single inotify fd?

**Comment by riccardomurri** [comment]
In cmdline.py:

>                 self.log.info("Running OI program")

Message content looks outdated; AFAICS, `oi` is no longer used.

**Comment by riccardomurri** [comment]
In cmdline.py:

>             lockfile = os.path.join(self.params.working_dir,
>                                     self.name + '.pid')

Wouldn't it make more sense to put the lockfile in the session
directory?  After all, we do not want to separate daemons managing the
same session; we do not care about them having the same working
directory, as long as the sessions and inboxes are separate...

**Comment by riccardomurri** [comment]
In cmdline.py:

>                 stdout=open(
>                     os.path.join(self.params.working_dir, 'stdout.txt'),
>                     'w'),
>                 stderr=open(
>                     os.path.join(self.params.working_dir, 'stderr.txt'),
>                     'w'),

These names should be configurable, including sending them to
`/dev/null` which is (AFAICT) what almost every daemon out there
does...

**Comment by riccardomurri** [comment]
Looks otherwise good to me :-)

File: gc3apps/irm.uzh.ch/gbeastcli.py (hunk 1 above)
Comment (arcimboldo) [inline gc3apps/irm.uzh.ch/gbeastcli.py]: No indeed, this script is not needed anymore. I've deleted it.

File: gc3libs/tests/scripts/daemoncli.py (hunk 6 above)
Comment (arcimboldo) [inline gc3libs/tests/scripts/daemoncli.py]: No. Removing old script.

**Comment by arcimboldo** [comment]
You need to keep a mapping between the directory and the inotify
descriptor, because events returned by `inotifyx.get_events` contain only
the *relative* path, and not the containing directory.

On Thu, Apr 21, 2016 at 11:31 PM, Riccardo Murri <notifications@github.com>
wrote:

> In cmdline.py:
>
> def __setup_inotify(self):
>     # Setup inotify on inbox directories
>     self.inotify_fds = {}
>     for inbox in self.params.inbox:
>         ifd = inotifyx.init()
>         self.inotify_fds[inbox] = ifd
>         inotifyx.add_watch(ifd, inbox, self.inotify_event_mask)
>
> Why is a different inotify object needed *per directory*?
> Can't you just add_watch() all directories on a single inotify fd?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/uzh/gc3pie/pull/536#issuecomment-213121481>
>



-- 
antonio.s.messina@gmail.com
antonio.messina@uzh.ch                     +41 (0)44 635 42 22
S3IT: Service and Support for Science IT   http://www.s3it.uzh.ch/
University of Zurich
Winterthurerstrasse 190
CH-8057 Zurich Switzerland

**Comment by arcimboldo** [comment]
In case you use --client, the script will fail if any non-compatible option
is used (actually: if all non compatible options have non-default values
set).

Although in principle I agree with your point, this should be discussed
further, because I don't think we could re-use the options set in
`_SessionBasedCommand`, so we either need to rewrite part of
_SessionBasedCommand and SessionBasedScript or we will need to replicate
common options like `-s` or `-r` in SessionBasedDaemon.


On Thu, Apr 21, 2016 at 11:14 PM, Riccardo Murri <notifications@github.com>
wrote:

> In cmdline.py:
>
>     self.add_param('--client',
>                    nargs='+',
>                    metavar="ARGS",
>
> I'm not sure I like this to be implemented as an *option*; the problem
> is that it is then legal to combine it with other options, and the
> result is (a priori) undefined: what should the effect of daemon
> --notify-state FOO --client BAR be?
>
> Wouldn't it be better to have the SessionDaemon class take a
> mandatory first argument: *server* or *client* and then the rest of
> the command line is parsed accordingly::
>
>     daemon server --notify-state FOO
>
>     daemon client /path/to/sock bar baz
>
> Doesn't solve the "grammar" problem (unless we switch, e.g., to click
> for parsing command-line options) but at least one can state in the
> help text "this option only works in *server* mode" or "this option
> only works in *client* mode".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/uzh/gc3pie/pull/536#issuecomment-213117028>
>



-- 
antonio.s.messina@gmail.com
antonio.messina@uzh.ch                     +41 (0)44 635 42 22
S3IT: Service and Support for Science IT   http://www.s3it.uzh.ch/
University of Zurich
Winterthurerstrasse 190
CH-8057 Zurich Switzerland

**Comment by arcimboldo** [comment]
There is, apparently, a race condition: when the thread is created, it's
likely that the session directory has not been created yet, thus the
lockfile will fail.

On Thu, Apr 21, 2016 at 11:35 PM, Riccardo Murri <notifications@github.com>
wrote:

> In cmdline.py:
>
>         lockfile = os.path.join(self.params.working_dir,
>                                 self.name + '.pid')
>
> Wouldn't it make more sense to put the lockfile in the session
> directory? After all, we do not want to separate daemons managing the
> same session; we do not care about them having the same working
> directory, as long as the sessions and inboxes are separate...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/uzh/gc3pie/pull/536#issuecomment-213122585>
>



-- 
antonio.s.messina@gmail.com
antonio.messina@uzh.ch                     +41 (0)44 635 42 22
S3IT: Service and Support for Science IT   http://www.s3it.uzh.ch/
University of Zurich
Winterthurerstrasse 190
CH-8057 Zurich Switzerland

**Comment by riccardomurri** [comment]
More comments and replies, in order of decreasing importance:

Tests are failing because apparently a directory is (re)used
across tests without being properly cleaned.  Please fix the
tests before merging.


> In case you use --client, the script will fail if any
> non-compatible option is used (actually: if all non compatible
> options have non-default values set).

I see, but my point is: this cannot be made not obvious to users
from the help/usage text -- whereas different command "verbs"
make it clear that they are alternative.

Anyway, I'm not going to insist much, as we have the same
situation on the SessionBasedScript.  However, it would be nice
not to accumulate more "debt" that might need to be fixed sooner
or later ;-)


> There is, apparently, a race condition: when the thread is
> created, it's likely that the session directory has not been
> created yet, thus the lockfile will fail.

Use a condition variable?


Anyway, looks generally good to me so feel free to merge as soon
as the tests are fixed.

**Comment by arcimboldo** [comment]
On Fri, Apr 22, 2016 at 10:11 PM, Riccardo Murri <notifications@github.com>
wrote:

> More comments and replies, in order of decreasing importance:
>
> Tests are failing because apparently a directory is (re)used
> across tests without being properly cleaned. Please fix the
> tests before merging.
>

travis seems happy now, is there something I'm missing?


> In case you use --client, the script will fail if any
> non-compatible option is used (actually: if all non compatible
> options have non-default values set).
>
> I see, but my point is: this cannot be made not obvious to users
> from the help/usage text -- whereas different command "verbs"
> make it clear that they are alternative.
>
> Anyway, I'm not going to insist much, as we have the same
> situation on the SessionBasedScript. However, it would be nice
> not to accumulate more "debt" that might need to be fixed sooner
> or later ;-)
>
If we move the setup of common options in an external function maybe we can
reuse it on a parser or a subparser, so that we could use it for both the
SessionBasedScript and the subparsers of the SessionBasedDaemon, what do
you think?


> There is, apparently, a race condition: when the thread is
> created, it's likely that the session directory has not been
> created yet, thus the lockfile will fail.
>
> Use a condition variable?
>
I actually think it make sense to put pidfile, logs and everything which is
specific to the daemon in a separate directory...


> Anyway, looks generally good to me so feel free to merge as soon
> as the tests are fixed.
>
>
I will do it tomorrow, unless you have other comments.
.a.

-- 
antonio.s.messina@gmail.com
antonio.messina@uzh.ch                     +41 (0)44 635 42 22
S3IT: Service and Support for Science IT   http://www.s3it.uzh.ch/
University of Zurich
Winterthurerstrasse 190
CH-8057 Zurich Switzerland

**Comment by riccardomurri** [comment]
> travis seems happy now, is there something I'm missing?

Hmmm, no ... I do not know what I was looking at yesterday, that had
issues in the tests, then...


> > In case you use --client, the script will fail if any
> > non-compatible option is used (actually: if all non compatible
> > options have non-default values set).
> >
> > I see, but my point is: this cannot be made not obvious to users
> > from the help/usage text -- whereas different command "verbs"
> > make it clear that they are alternative.
> >
> > Anyway, I'm not going to insist much, as we have the same
> > situation on the SessionBasedScript. However, it would be nice
> > not to accumulate more "debt" that might need to be fixed sooner
> > or later ;-)
> >
> If we move the setup of common options in an external function maybe we can
> reuse it on a parser or a subparser, so that we could use it for both the
> SessionBasedScript and the subparsers of the SessionBasedDaemon, what do
> you think?

Let's talk about this, but it'd be a topic for another feature branch /
pull request.


> > There is, apparently, a race condition: when the thread is
> > created, it's likely that the session directory has not been
> > created yet, thus the lockfile will fail.
> >
> > Use a condition variable?
> >
> I actually think it make sense to put pidfile, logs and everything which is
> specific to the daemon in a separate directory...

My point is: I would like to avoid the case where two independent daemon
instances operate on the same session.  Whether you do this by a
lockfile in the session directory or some more sophisticated way, I do
not care much.

Ciao,
R

--
Riccardo Murri, Anna-Heer-Strasse 10, CH-8057 Zürich, Switzerland

**Comment by arcimboldo** [comment]
On Sat, Apr 23, 2016 at 9:44 AM, Riccardo Murri <notifications@github.com>
wrote:

> > > In case you use --client, the script will fail if any
> > > non-compatible option is used (actually: if all non compatible
> > > options have non-default values set).
> > >
> > > I see, but my point is: this cannot be made not obvious to users
> > > from the help/usage text -- whereas different command "verbs"
> > > make it clear that they are alternative.
> > >
> > > Anyway, I'm not going to insist much, as we have the same
> > > situation on the SessionBasedScript. However, it would be nice
> > > not to accumulate more "debt" that might need to be fixed sooner
> > > or later ;-)
> > >
> > If we move the setup of common options in an external function maybe we
> can
> > reuse it on a parser or a subparser, so that we could use it for both the
> > SessionBasedScript and the subparsers of the SessionBasedDaemon, what do
> > you think?
>
> Let's talk about this, but it'd be a topic for another feature branch /
> pull request.
>
>
>
Too bad, because I just implemented it. However, it's a bit ugly...


> > > There is, apparently, a race condition: when the thread is
> > > created, it's likely that the session directory has not been
> > > created yet, thus the lockfile will fail.
> > >
> > > Use a condition variable?
> > >
> > I actually think it make sense to put pidfile, logs and everything which
> is
> > specific to the daemon in a separate directory...
>
> My point is: I would like to avoid the case where two independent daemon
> instances operate on the same session. Whether you do this by a
> lockfile in the session directory or some more sophisticated way, I do
> not care much.
>

Wouldn't it make sense to always add a lockfile to the session dir? You
have the same problem when running  a script with -C and run the same
script on another shell, or the g* commands.

However, to simplify things there is the --working-dir option. If you run
the daemon with --working-dir  only the session is created in the working
directory, and since the pidfile is locked you cannot run the same daemon
twice with the same working directory.

.a.

-- 
antonio.s.messina@gmail.com
antonio.messina@uzh.ch                     +41 (0)44 635 42 22
S3IT: Service and Support for Science IT   http://www.s3it.uzh.ch/
University of Zurich
Winterthurerstrasse 190
CH-8057 Zurich Switzerland
